### PR TITLE
periodicity of e**(I*x)

### PR DESCRIPTION
**Pull request by friyaz**
#### Brief description of what is fixed or changed
`periodicity()` can now calculate period of  expressions of form  `exp(I*x)` when x is `real`.
`None` is returned for complex `x`.

```
>>> from sympy import *
>>> from sympy.calculus.util import periodicity
>>> x = Symbol('x', real=True)
>>> y = Symbol('y')
>>> periodicity(exp(I*x))
2*pi 
>>> periodicity(exp(y))
```

#### Release Notes

* calculus
    * fix periodicity of exponential of form exp(I*x)

### CODE REVIEW

**Comment by smichr** [comment]
looks ok

**Comment by oscarbenjamin** [comment]
Thanks for looking at this.

At first when looking at this I though it was maybe incorrect because it only checks the real part. Let me think it through out loud...

The main change is this:
```
    if isinstance(f, exp):
        period = periodicity(re(f), symbol)
```
So if I have `f(x) = exp(g(x) + I*h(x))` with g, h real then the `re(f(x)) = exp(g(x))*cos(h(x))` and `im(f(x)) = exp(g(x))*sin(h(x))`. Then cos and sin have the same periodicity so `cos(h(x))` and `sin(h(x))` have the same periodicity. That means `re(f(x))` has the same periodicity as `im(f(x))` so `f(x)` has the same periodicity as `re(f(x))`.

Am I missing anything?

**Comment by jksuom** [comment]
I think that the period of the real part should be checked on the original function.

**Comment by oscarbenjamin** [comment]
> I think that the period of the real part should be checked on the original function.

I don't understand this comment @jksuom.

That sounds like a description of what the PR already does but it also seems like you're suggesting that the PR should do something different.

**Comment by FrackeR011** [comment]
It went into infinite recursion for exp(x) when `x` is real. Now it checks periodicity only when argument of exp is complex.

**Comment by jksuom** [comment]
The real part of a function may be periodic but the imaginary part may be non-periodic. So it should be checked that a period of the real part is also a period of the imaginary part or, what amounts to the same, a period of the whole function.

**Comment by oscarbenjamin** [comment]
@jksuom can you give an example?

If the real part is `exp(g(x))*cos(h(x))` and the imaginary part is `exp(g(x))*sin(h(x))` how can they have different periodicity?

**Comment by jksuom** [comment]
Let `f` be a multivalued function that is branched a 0 and whose branches differ by an imaginary constant, for example, an exponential integral. Then the real part of `f(exp(I*x))` has period `2*pi` but the imaginary part is non-periodic.

**Comment by oscarbenjamin** [comment]
The change suggested in this PR is that for any function of the form `exp(f(x)))` we should have `periodicity(exp(f(x))) == periodicity(re(exp(f(x))))`. (unless I'm misunderstanding)

An example that breaks this could be:
```julia
In [10]: exp(log(sin(x)+I*cos(2*x)))                                                                                              
Out[10]: sin(x) + ⅈ⋅cos(2⋅x)
```
But this code will only kick in if it is an exp so we would need:
```julia
In [11]: exp(log(sin(x)+I*cos(2*x)), evaluate=False)                                                                              
Out[11]: 
 log(sin(x) + ⅈ⋅cos(2⋅x))
ℯ      
```

But then the real part doesn't work properly anyway:
```julia
In [12]: re(exp(log(sin(x)+I*cos(2*x)), evaluate=False))                                                                          
Out[12]: cos(arg(sin(x) + ⅈ⋅cos(2⋅x)))⋅│sin(x) + ⅈ⋅cos(2⋅x)│

In [13]: x = Symbol('x', real=True)                                                                                               

In [14]: re(exp(log(sin(x)+I*cos(2*x)), evaluate=False))                                                                          
Out[14]: 
   _____________________                              
  ╱    2         2                                    
╲╱  sin (x) + cos (2⋅x) ⋅cos(arg(sin(x) + ⅈ⋅cos(2⋅x)))
```

**Comment by jksuom** [comment]
In general, if the real part of `f` has period `a`, then the real part of `F(x) = f(x + a) - f(x)`  will be identically zero. If `f` and `F` are analytic (as we usually expect), then `F` must have an imaginary constant value `c` that may be nonzero. Then `f` will satisfy `f(x + a) = f(x) + c`, i.e., it is quasiperiodic.

**Comment by oscarbenjamin** [comment]
What you're saying makes sense but we need test case examples for this PR.

**Comment by FrackeR011** [comment]
> The change suggested in this PR is that for any function of the form `exp(f(x)))` we should have `periodicity(exp(f(x))) == periodicity(re(exp(f(x))))`
This was failing for exp(x) when x was real as it was re(exp(x) equals exp(x).

So i have changed it to work only when exp(f(x)) has some non-zero imaginary part.
Then claim here is
`periodicity(exp(f(x)) == periodicity(im(exp(f(x)))) `     
only if im(exp(fx)) == 0 

I tried 
```
>>> x = Symbol(x, real=True)
>>> t = exp(log(sin(x)+I*cos(2*x)),evaluate= False)
>>> periodicitiy(t,x)
2*pi
```
So it works in this case.

> What you're saying makes sense but we need test case examples for this PR.
OK. I will add more test cases.

**Comment by oscarbenjamin** [comment]
Maybe a better way to do this is by rewriting `exp`:
```julia
In [28]: exp(I*x).rewrite(cos)                                                                                                    
Out[28]: ⅈ⋅sin(x) + cos(x)
```

However that doesn't work
```julia
In [27]: print(periodicity(cos(x) + I*sin(x), x))                                                                                 
None
```
so that should also be fixed.

**Comment by FrackeR011** [comment]
> Maybe a better way to do this is by rewriting `exp`
That does not test the condition as expression will not remain in the form `exp(f(x))`

> However that doesn't work
It works if x is real.

Do you know why Travis Build is failing? I do not think it is related to this fix.

**Comment by oscarbenjamin** [comment]
Does `x` need to be real for this?

It doesn't need to be real in this case:
```julia
In [18]: x = Symbol('x')                                                                                                          

In [19]: periodicity(cos(x), x)                                                                                                   
Out[19]: 2⋅π
```

So why this:
```julia
In [20]: periodicity(cos(x) + I*sin(x), x)                                                                                        

```

File: sympy/calculus/tests/test_util.py [hunk 1]
```diff
@@ -122,6 +123,15 @@ def test_periodicity():
     assert periodicity(tan(x), y) == S.Zero
 
     assert periodicity(exp(x), x) is None
+    assert periodicity(exp(I*x), x) is None
+    assert periodicity(exp(I*z), z) == 2*pi
+    assert periodicity(exp(z), z) is None
+    assert periodicity(exp(log(sin(z) + I*cos(2*z)), evaluate=True), z) == 2*pi
+    assert periodicity(exp(log(sin(2*z) + I*cos(z)), evaluate=True), z) == 2*pi
```
Comment (asmeurer) [inline sympy/calculus/tests/test_util.py]: I think these were meant to be `evaluate=False`. `evaluate=True` is already the default.

**Comment by asmeurer** [comment]
This looks good to me except for the one comment. I agree that if the function is an exponential with nonzero imaginary part the periodicity of the imaginary part should equal the periodicity of the real part, so it is only necessary to check one. For larger expressions that won't be true, but my understanding is that `periodicity` works recursively

**Comment by asmeurer** [comment]
Ah I see, `cos(h)` and `sin(h)` can have different periodicity. For example, if `h` is `arg(sin(x) + I*cos(2*x))`. Then

```py
>>> pprint(cos(arg(sin(x) + I*cos(2*x)).rewrite(atan2)))
         sin(x)
────────────────────────
   _____________________
  ╱    2         2
╲╱  sin (x) + cos (2⋅x)
>>> pprint(sin(arg(sin(x) + I*cos(2*x)).rewrite(atan2)))
        cos(2⋅x)
────────────────────────
   _____________________
  ╱    2         2
╲╱  sin (x) + cos (2⋅x)
```

So I agree we need to test both `re(f)` and `im(f)`. 

This also means this is wrong:

```
>>> periodicity(im(exp(log(sin(z) + I*cos(2*z)), evaluate=False)), z)
2*pi
```

and therefore we need a better test.

**Comment by FrackeR011** [comment]
> So I agree we need to test both `re(f)` and `im(f)`

Nice catch. I will change it to `lcim` of period of real and imaginary part.

periodicity is giving wrong result here:
```
>>> x = Symbol('x', real=True)
>>> h = sin(arg(sin(x) + I*cos(2*x)).rewrite(atan2))
>>> periodicity(h, x)
2*pi
```
Answer should be pi.

**Comment by FrackeR011** [comment]
> This also means this is wrong:

```
>>> periodicity(im(exp(log(sin(z) + I*cos(2*z)), evaluate=False)), z)
2*pi
```

> and therefore we need a better test.

Here we are checking periodicity with respect to argument of sine for which 2*pi seems to be correct answer. Can you explain why it is wrong?

**Comment by asmeurer** [comment]
The imaginary part is `cos(2*z)`. You can see if you remove the evaluate=False it returns `pi` (you also have to make `z` real to prevent `im(cos(2*z))` from evaluating into sinh and cosh).

**Comment by asmeurer** [comment]
To be sure, the period is correct, it just isn't the fundamental period. The `periodicity` docstring does indicate this can happen. If you replace `cos(2*z)` with a nonperiodic function, it returns `None`, so it doesn't actually give a truly wrong result in any examples I tested.

**Comment by oscarbenjamin** [comment]
Thanks, that looks better.

We still don't get a result for `periodicity(cos(x) + I*sin(x), x)` even though we do for `cos(x)` and `sin(x)` and this PR only works for `exp(I*x)` if `x` is real but it shouldn't need to be real.

The [docstring](https://github.com/sympy/sympy/blob/b68caad822d3ad0ae1fa09c78c9ed2db927e73cb/sympy/calculus/util.py#L353) says that
```
    Currently, we do not support functions with a complex period.
    The period of functions having complex periodic values such
    as `exp`, `sinh` is evaluated to `None`.
```
but I don't understand this. The conditions on `x` such that `cos(x)` has period `2*pi` are no different from the case of `exp(I*x)`.

**Comment by asmeurer** [comment]
I think complex period specifically means the result of periodicity would be a complex (nonreal) number. `exp(I*x)` takes on complex values but is periodic on the real line. 

I agree `x` shouldn't need to be set as real. It probably gets confused by `im(x)` and `re(x)`. If `periodicity` implicitly assumes the variable is real, then `periodicity(im(x), x)` should be the same as `periodicity(0, x) == 0`.

**Comment by FrackeR011** [comment]
Let `x = a + I*b`. Then i think period of exp(x) should be `I*2*pi` which is complex.
```
>>> a, b =symbols('a b', real=True)
>>> x = a + I*b
>>> exp(x) == exp(x + I*2*pi)
True
>>> exp(x) == exp(x + 2*pi)
False
```
Maybe something similar happens for `periodicity(cos(x) + I*sin(x), x)` when x is complex.

**Comment by oscarbenjamin** [comment]
It works fine if the I is replaced by something else:
```julia
In [42]: periodicity(cos(x)+I*sin(x), x)                                                                                          

In [43]: periodicity(cos(x)+y*sin(x), x)                                                                                          
Out[43]: 2⋅π
```
Maybe there's an explicit check for `I` somewhere...

**Comment by FrackeR011** [comment]
`None` is returned when calculating  period of cos(re(x)) with respect to x. This leads to None being returned for `exp(I*x).

```
>>> x = symbols('x')
>>> r = re(exp(I*x))
>>> r
exp(-im(x))*cos(re(x))
>>> periodicity(r, x)
None
>>>
>>> periodicity(cos(re(x)),  x)
>>>
```

**Comment by asmeurer** [comment]
Most likely a check that replaces `re(x)` with `x` and `im(x)` with `0` should do it. We could also swap out `x` for a real Dummy. If we ever want to support complex periodicity the logic would need to be better, but as long as we implicitly assume the variable is real, we might as well force it to be so.

**Comment by oscarbenjamin** [comment]
What about just putting `f.subs(Symbol('x', real=True))` right at the top of the function?

**Comment by asmeurer** [comment]
That's what I meant by swapping out x for a real Dummy. That's probably better than handling `re` and `im`, as there might be other functions that automatically reduce on real inputs, and it would allow the assumptions to propagate up.

**Comment by FrackeR011** [comment]
@asmeurer @oscarbenjamin Please review.

**Comment by oscarbenjamin** [comment]
Looks good except that you should use `Dummy` rather than symbol. Otherwise there is a possibility that the symbol you use might clash with another symbol in the input:
```julia
In [16]: x = Symbol('x', real=True)                                                                                                                            

In [17]: print(periodicity(x-sin(y), y))  
None
```

**Comment by oscarbenjamin** [comment]
This looks good. Thanks for this!

**Pull request merged**